### PR TITLE
Remove the concept of packable fields from EEClass

### PR DESCRIPTION
**Pull request by davidwrighton**
- EEClass has long had a concept of packable rarely accessed fields
- When we had NGEN we had logic to make these fields take very small amounts of space, but now that we no longer have NGEN, this code is not actually used in practice
- The unpacked form of these fields used 4 bytes per field, even though many of them were logically 2 byte fields
- The NumBoxed fields were just unnecessary, especially if we have a HasBoxedThreadStatics flag, which this change adds
- Move the HasBoxed flags to the WFLAGS_LOW_ENUM

Overall this saves 24 bytes per EEClass and moves all these fields onto a normal structure.

### CODE REVIEW

File: src/coreclr/vm/array.cpp [hunk 1]
```diff
@@ -324,7 +324,7 @@ MethodTable* Module::CreateArrayMethodTable(TypeHandle elemTypeHnd, CorElementTy
 
         // Allocate ArrayClass (including space for packed fields), MethodTable, and class name in one alloc.
```
Comment (AustinWise) [inline src/coreclr/vm/array.cpp:325]:
```suggestion
        // Allocate ArrayClass, MethodTable, and class name in one alloc.
```

File: src/coreclr/vm/class.h [hunk 2]
```diff
@@ -1888,7 +1803,10 @@ class EEClass // DO NOT CREATE A NEW EEClass USING NEW!
     /*
      * Constructor: prevent any other class from doing a new()
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/class.h:1804]: Except for subclasses?

File: src/coreclr/vm/methodtablebuilder.cpp [hunk 3]
```diff
@@ -115,7 +115,7 @@ MethodTableBuilder::CreateClass( Module *pModule,
     }
     else
     {
-        pEEClass = new (pAllocator->GetLowFrequencyHeap(), pamTracker) EEClass(sizeof(EEClass));
+        pEEClass = new (pAllocator->GetLowFrequencyHeap(), pamTracker) EEClass();
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/methodtablebuilder.cpp:118]: There are a few comments about not instantiating an EEClass with `new`, looks like we can remove or update them?

Comment (davidwrighton) [reply src/coreclr/vm/methodtablebuilder.cpp:118]: Eh, `EEClass` still doesn't allow for the normal boring new operator. For correct usage, you still need the more complex new expression here. We could remove it, and require placement new in into a LoaderHeap in all scenarios, which would also work, but I don't see a need to remove the weird new behavior at this time.

File: src/coreclr/vm/class.cpp [hunk 4]
```diff
@@ -209,7 +190,7 @@ EEClass::CreateMinimalClass(LoaderHeap *pHeap, AllocMemTracker *pamTracker)
     }
     CONTRACTL_END;
 
-    return new (pHeap, pamTracker) EEClass(sizeof(EEClass));
+    return new (pHeap, pamTracker) EEClass();
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/coreclr/vm/class.cpp:193]: We still want to allocate using the `LoaderHeap` as that is what controls object lifetime. We have no mechanism in place to run delete and free memory for these piecemeal.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**